encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/ReflectiveObjectDecoderProvider.java

```diff
@@ -0,0 +1,210 @@
+      while (currentClass != Object.class && currentClass != null) {
+        for (Field field : currentClass.getDeclaredFields()) {
+          if (!shouldIncludeField(field)) {
+            continue;
+          }
+          String fieldName = field.getName();
+          if (descriptors.get(fieldName) == null) {
+            descriptors.put(fieldName, buildFieldDescriptor(field));
+          }
+          if (reflectiveSetters.get(fieldName) == null) {
+            reflectiveSetters.put(fieldName, ReflectiveFieldSetter.of(field));
+          }
+        }
+        currentClass = currentClass.getSuperclass();
+      }
+    }
+
+    private void decodeFields(ObjectDecoderContext<T> ctx) {
+      for (Map.Entry<String, ReflectiveSetter> entry : reflectiveSetters.entrySet()) {
+        String fieldName = entry.getKey();
+        Class<?> fieldType = entry.getValue().getFieldRawType();
+        FieldDescriptor fieldDescriptor = descriptors.get(fieldName);
+        if (fieldDescriptor == null) {
+          throw new RuntimeException(fieldName + " did not have a FieldDescriptor.");
+        }
+        FieldRef<?> ref;
+        if (fieldType.equals(int.class)) {
+          ref = ctx.decodeInteger(fieldDescriptor);
+        } else if (fieldType.equals(long.class)) {
+          ref = ctx.decodeLong(fieldDescriptor);
+        } else if (fieldType.equals(short.class)) {
+          ref = ctx.decodeShort(fieldDescriptor);
+        } else if (fieldType.equals(double.class)) {
+          ref = ctx.decodeDouble(fieldDescriptor);
+        } else if (fieldType.equals(float.class)) {
+          ref = ctx.decodeFloat(fieldDescriptor);
+        } else if (fieldType.equals(boolean.class)) {
+          ref = ctx.decodeBoolean(fieldDescriptor);
+        } else {
+          TypeToken<?> fieldTypeToken =
+              getFieldTypeToken(entry.getValue().getFieldGenericType(), ctx);
+          if (entry.getValue().isDecodedInline()) {
+            if (fieldTypeToken instanceof TypeToken.ClassToken) {
+              @SuppressWarnings("unchecked")
+              TypeToken.ClassToken<Object> classToken =
+                  (TypeToken.ClassToken<Object>) fieldTypeToken;
+              ref = ctx.decodeInline(classToken, ReflectiveObjectDecoder.DEFAULT);
+            } else {
+              throw new IllegalArgumentException(
+                  "Array types cannot be decoded inline, type:" + fieldTypeToken + " found.");
+            }
+          } else {
+            ref = ctx.decode(fieldDescriptor, fieldTypeToken);
+          }
+        }
+        refs.put(fieldName, ref);
+      }
+    }
+
+    private TypeToken<?> getFieldTypeToken(Type type, ObjectDecoderContext<?> ctx) {
+      if (type instanceof TypeVariable) {
+        TypeVariable[] typeVariables = ctx.getTypeToken().getRawType().getTypeParameters();
+        for (int i = 0; i < typeVariables.length; i++) {
+          if (typeVariables[i].equals(type)) {
+            return ctx.getTypeArgument(i);
```

Comment:
don't we want to throw a meaningful exception in the case we could not find the TypeVariable in the class definition? By "meaningful" I mean, not the generic on that `TypeToken.of` will throw, but rather something along the lines of "class foo<T,U,V> does not have type variable W".